mcx/diagnostics/gelman_rubin.py

```diff
@@ -1,38 +1,156 @@
 """Kernel to compute the Gelman-Rubin convergence diagnostic (Rhat) online.
 """
-from typing import NamedTuple
+from typing import Callable, NamedTuple, Tuple
 
+import jax
 import jax.numpy as jnp
 
-from mcx.inference.warmup.mass_matrix_adaptation import (
-    WelfordAlgorithmState,
-    welford_algorithm,
-)
+
+class WelfordAlgorithmState(NamedTuple):
```

Comment:
I didnt realize you could set up a namedtuple constructor like this. This is amazing

Reply:
I'm addicted to those. Working around lightweight data structure usually
leads to a codebase that is easier to understand.

Reply:
Readability is +++!
Though I do remember David Beazley posting something on twitter on dataclasses being slow.

Reply:
As with everything, it depends. When you unpack a NamedTuple as a tuple it is as fast as it gets, and it is slightly slower when you access variables like `namedtuple.var`. Dataclasses are slightly faster than NamedTuples when you do `dataclass.var`, but objects are slower to create. Having said that they're all in the same ballpark, and we are limited here by computation of the gradients, not IO :grin: 

The only reason I chose NamedTuple over Dataclasses is that they play nicer with JAX out of the box.